internal/webserver/metadatatasks/taskpool.go

```diff
@@ -0,0 +1,62 @@
+package metadatatasks
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/SwissOpenEM/Ingestor/internal/metadataextractor"
+)
+
+type MetadataExtractionTaskPool struct {
+	tasks   chan task
+	wg      sync.WaitGroup
+	handler *metadataextractor.ExtractorHandler
+}
+
+func (p *MetadataExtractionTaskPool) GetAvailableMethods() []metadataextractor.MethodAndSchema {
+	return p.handler.AvailableMethods()
+}
+
+func (p *MetadataExtractionTaskPool) NewTask(ctx context.Context, datasetPath string, method string) (*ExtractionProgress, error) {
+	epc := ExtractionProgress{
+		ProgressSignal: make(chan bool, 1),
+	}
+
+	select {
+	case p.tasks <- task{
+		ctx:          ctx,
+		datasetPath:  datasetPath,
+		method:       method,
+		taskProgress: &epc,
+	}:
+		return &epc, nil
+	default:
+		return nil, errors.New("task queue is full")
```

Comment:
In case there are too many requests, the user either
1) gets this error in the request and needs to try again
2) well get a timeout in the request if the queue size is not limited and the extraction takes too long to be started/executed

Is my understanding correct? Or is 2) not possible because of the SSE implementation?
Is the queue limit to prevent users from spamming the service? (that is probably a good idea ^^)

Reply:
The queue limit comes from the buffered channel setup which has a limited number of slots (we don't have infinite RAM). It also works in FIFO mode meaning it keeps the request order. 

Unbuffered channels (the default for channels) will block until someone consumes your value. You can use a "select" block to get around it, but this will definitely not preserve the "first come, first serve" order of the buffered method. Additionally, If we keep it blocked until consumed, then we can't update the user about the current status in SSE in the mean time.

If we reach the queue limit, then an error is thrown by the section you highlighted. However, this is handled by the controller in the function passed as the "step" argument in `g.Stream` (line 38 in `webserver/metadata.go`). The step function uses `g.SSEvent` to create SSE messages, and returns `false` to indicate that we're not done. After the g.Stream streams the data, it calls the step function again in this case. (otherwise, it will end the stream) 

This "step" function is where we send SSE updates, and the function works as a basic state machine: 
1. attempt to insert into queue -> if fails (the error above is thrown), update client and retry in 1 minute; on success, go to next state
2. wait for worker to consume our request -> if not consumed in 1 minute, update client, then continue waiting; if consumed, go to next state
3. wait worker to finish, and send all updates created by the worker/extractor. When the extractor finishes, send final update and end the SSE stream

Reply:
All this means is that if the user spams requests, it will not get straight up rejected, just blocked until their request can go into the queue (or if they drop the connection, then their request will not be handled further)

Reply:
Thanks for the explanation!